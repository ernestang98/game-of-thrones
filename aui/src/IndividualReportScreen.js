import React, { useEffect } from 'react';
import {
    Typography,
    Box,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
    Stack,
    List,
    ListItem,
    ListItemText,
    ListItemIcon,
    Fab
} from "@mui/material";
import {Link} from "react-router-dom";
import MaterialTable from '@material-table/core'
import axios from 'axios';

import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import { forwardRef } from 'react';
import Check from '@material-ui/icons/Check';
import ChevronLeft from '@material-ui/icons/ChevronLeft';
import ChevronRight from '@material-ui/icons/ChevronRight';
import Clear from '@material-ui/icons/Clear';
import DeleteOutline from '@material-ui/icons/DeleteOutline';
import Edit from '@material-ui/icons/Edit';
import FilterList from '@material-ui/icons/FilterList';
import FirstPage from '@material-ui/icons/FirstPage';
import LastPage from '@material-ui/icons/LastPage';
import Remove from '@material-ui/icons/Remove';
import SaveAlt from '@material-ui/icons/SaveAlt';
import Search from '@material-ui/icons/Search';
import ViewColumn from '@material-ui/icons/ViewColumn';

import FormatListNumberedIcon from '@mui/icons-material/FormatListNumbered';
import MovingIcon from '@mui/icons-material/Moving';
import ShieldIcon from '@mui/icons-material/Shield';
import NavigationIcon from '@mui/icons-material/Navigation';

/**
 * render Individual Report Screen
 * @returns {JSX.Element}
 * @constructor
 */
function IndivReportScreen() {

    const tableIcons = {
        Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
        Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
        Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
        Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
        DetailPanel: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
        Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
        Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
        Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
        FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
        LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
        NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
        PreviousPage: forwardRef((props, ref) => <ChevronLeft {...props} ref={ref} />),
        ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
        Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
        SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
        ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
        ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />)
    };

    const std = [

    ]
    const [students, setStudents] = React.useState(std);

    /**
     * Retrieve all students to display in dropdown select
     */
    const getAllStudents = () => {
        const endpoint = process.env.REACT_APP_API_ENDPOINT === null ? "https://fastapi-ernestang98.cloud.okteto.net/" : process.env.REACT_APP_API_ENDPOINT
        axios.get(endpoint + "student/")
        .then(function (res) {
            // console.log(res);
            setStudents(res.data);
            // console.log(students);
        })
        .catch(function (error) {
            console.log(error);
        })
    }

    const [isSelected, setSelected] = React.useState(false);
    const [selectedStudent, setSelectedStudent] = React.useState({
        /**
         * Student email
         */
        email: "",
        /**
         * Student's total points earned so far
         */
        points: 0,
        /**
         * Student's in-game rank
         */
        rank: 0,
        /**
         * Student's in-game position
         */
        position: 0
    });

    const [studentReport, setStudentReport] = React.useState([]);

    /**
     * Everytime admin selects a new student, set state of student details
     * @param {String} event 
     */
    const handleSelect = (event) => {
        let studentEmail= event.target.value;
        setSelected(true);
        console.log(studentEmail);
        getStudentDetails(studentEmail);
    };
    
    /**
     * Retrieve selected student's specific scores, rank, position etc
     * @param {String} studentEmail 
     */
    const getStudentDetails = (studentEmail) => {
        const endpoint = process.env.REACT_APP_API_ENDPOINT === null ? "https://fastapi-ernestang98.cloud.okteto.net/" : process.env.REACT_APP_API_ENDPOINT
        axios.all([
            axios.get(endpoint + "student/" + studentEmail),
            axios.get(endpoint + "attempt/", { params: { student_email: studentEmail } }),
            axios.get(endpoint + "attempt/cohortresults/", { params: { student_email: studentEmail } })
        ])
        .then(axios.spread((obj1, obj2, obj3) => {
            console.log(obj3.data[0]);
            if (typeof obj3.data[0] != 'undefined') {
                var totalCorrectPerc = ((obj3.data[0].points_earned / obj3.data[0].max_points_earnable)*100).toFixed(2);
                setSelectedStudent(prevState => ({
                    ...prevState,
                    email: obj1.data.email,
                    totalCorrectPerc: totalCorrectPerc,
                    rank: obj1.data.rank,
                    position: obj1.data.position
                }));

                let reportArray = obj2.data;
                reportArray.forEach( function(arrayItem) {
                    arrayItem["percentage_correct"] = (( arrayItem["points_scored"] / arrayItem["total_points"] ) * 100).toFixed(2);
                })
                setStudentReport(reportArray);
            } else {
                setSelectedStudent({});
                setStudentReport([]);
            }
        }))
    }

    /**
     * Hook to retrieve student results before rendering of DOM
     */
    useEffect( ()=> {
        getAllStudents();
    })

    return (
            <Box sx={{ my: 5, justifyContent: "center", flexDirection: "row", alignItems: "center" }}>
                <Stack
                spacing={5}
                justifyContent="center"
                alignItems="center">
                    <Fab id="cohort-link" variant="extended" color="primary" component={Link} to={"/cohort-report"}>
                        <NavigationIcon sx={{ mr: 1 }} />
                        Cohort Report
                    </Fab>
                    <Typography align="center" variant="h3">
                        Select Student
                    </Typography>
                    <Box
                    sx={{
                        width: 300,
                        // height: 300,
                    }}
                    >
                        <FormControl fullWidth>
                            <InputLabel id="demo-simple-select-label">Name</InputLabel>
                            <Select
                                labelId="demo-simple-select-label"
                                id="demo-simple-select"
                                onChange={handleSelect}
                                label="Name"
                                defaultValue=''
                            >
                                {students.map((student) => {
                                    return (
                                        <MenuItem key={student.name} value={student.email}>{student.name}</MenuItem>
                                    )
                                })}
                            </Select>
                        </FormControl>
                        <Box sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}>
                            <List>
                                <ListItem>
                                    <ListItemIcon>
                                        <MovingIcon />
                                    </ListItemIcon>
                                    <ListItemText>
                                        Overall Correct: {selectedStudent.totalCorrectPerc}%
                                    </ListItemText>
                                </ListItem>
                                <ListItem>
                                    <ListItemIcon>
                                        <ShieldIcon />
                                    </ListItemIcon>
                                    <ListItemText>
                                        Rank: {selectedStudent.rank}
                                    </ListItemText>
                                </ListItem>
                                <ListItem>
                                    <ListItemIcon>
                                        <FormatListNumberedIcon />
                                    </ListItemIcon>
                                    <ListItemText>
                                        Position: {selectedStudent.position}
                                    </ListItemText>
                                </ListItem>
                            </List>
                        </Box>
                    </Box>
                    <Box
                    sx={{
                        width: "100%",
                        height: 300,
                    }}
                    >
                        <MaterialTable

                            columns={[
                                { title: 'Quest Name', field: 'quest_name' },
                                { title: 'Points Scored', field: 'points_scored', type: 'numeric' },
                                { title: 'Total Possible Points', field: 'total_points', type: 'numeric' },
                                { title: '% Correct', field: 'percentage_correct', type: 'numeric' },
                                { title: 'Time to Complete', field: 'time_to_complete_in_seconds', type: 'numeric' },
                                { title: 'Total Possible Points', field: 'total_points', type: 'numeric' },
                            ]}
                            data={studentReport}
                            title="Report"
                            options={{filtering:true}}
                            icons={tableIcons}


                        />
                    </Box>
                </Stack>
            </Box>

    );
}
export default IndivReportScreen
