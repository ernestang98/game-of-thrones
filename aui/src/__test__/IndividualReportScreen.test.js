import React from 'react';
import {mount, shallow} from 'enzyme';
import IndividualReportScreen from "../IndividualReportScreen";

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import "@testing-library/jest-dom/extend-expect"
import toJson from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });

describe("Individual Report Screen", () => {
    it("render success", () => {
        const wrapper = shallow(<IndividualReportScreen />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('link to cohort report screen', () => {
        const wrapper = shallow(<IndividualReportScreen />);
        expect(wrapper
            .find('#cohort-link')
            .props()
            .to
            )
            .toBe('/cohort-report');
    })

    it("should update selected student state on select student", () => {
        const selectStudent = jest.fn();
        const wrapper = shallow(<IndividualReportScreen />);
        const handleClick = jest.spyOn(React, "useState");
        handleClick.mockImplementation(selectedStudent => [selectedStudent, setSelectedStudent]);
     
        wrapper.find("#demo-simple-select").simulate("click");
        expect(selectStudent).toBeTruthy();
    });
});

