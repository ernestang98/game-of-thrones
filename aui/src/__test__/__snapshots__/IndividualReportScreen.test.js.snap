// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Individual Report Screen render success 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "my": 5,
    }
  }
>
  <ForwardRef(Stack)
    alignItems="center"
    justifyContent="center"
    spacing={5}
  >
    <ForwardRef(Fab)
      color="primary"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      id="cohort-link"
      to="/cohort-report"
      variant="extended"
    >
      <Memo(ForwardRef(NavigationIcon))
        sx={
          Object {
            "mr": 1,
          }
        }
      />
      Cohort Report
    </ForwardRef(Fab)>
    <ForwardRef(Typography)
      align="center"
      variant="h3"
    >
      Select Student
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      sx={
        Object {
          "width": 300,
        }
      }
    >
      <ForwardRef(FormControl)
        fullWidth={true}
      >
        <ForwardRef(InputLabel)
          id="demo-simple-select-label"
        >
          Name
        </ForwardRef(InputLabel)>
        <ForwardRef(Select)
          defaultValue=""
          id="demo-simple-select"
          label="Name"
          labelId="demo-simple-select-label"
          onChange={[Function]}
        />
      </ForwardRef(FormControl)>
      <ForwardRef(Box)
        sx={
          Object {
            "bgcolor": "background.paper",
            "maxWidth": 360,
            "width": "100%",
          }
        }
      >
        <ForwardRef(List)>
          <ForwardRef(ListItem)>
            <ForwardRef(ListItemIcon)>
              <Memo(ForwardRef(MovingIcon)) />
            </ForwardRef(ListItemIcon)>
            <ForwardRef(ListItemText)>
              Overall Correct: 
              %
            </ForwardRef(ListItemText)>
          </ForwardRef(ListItem)>
          <ForwardRef(ListItem)>
            <ForwardRef(ListItemIcon)>
              <Memo(ForwardRef(ShieldIcon)) />
            </ForwardRef(ListItemIcon)>
            <ForwardRef(ListItemText)>
              Rank: 
              0
            </ForwardRef(ListItemText)>
          </ForwardRef(ListItem)>
          <ForwardRef(ListItem)>
            <ForwardRef(ListItemIcon)>
              <Memo(ForwardRef(FormatListNumberedIcon)) />
            </ForwardRef(ListItemIcon)>
            <ForwardRef(ListItemText)>
              Position: 
              0
            </ForwardRef(ListItemText)>
          </ForwardRef(ListItem)>
        </ForwardRef(List)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
    <ForwardRef(Box)
      sx={
        Object {
          "height": 300,
          "width": "100%",
        }
      }
    >
      <Table
        columns={
          Array [
            Object {
              "field": "quest_name",
              "title": "Quest Name",
            },
            Object {
              "field": "points_scored",
              "title": "Points Scored",
              "type": "numeric",
            },
            Object {
              "field": "total_points",
              "title": "Total Possible Points",
              "type": "numeric",
            },
            Object {
              "field": "percentage_correct",
              "title": "% Correct",
              "type": "numeric",
            },
            Object {
              "field": "time_to_complete_in_seconds",
              "title": "Time to Complete",
              "type": "numeric",
            },
            Object {
              "field": "total_points",
              "title": "Total Possible Points",
              "type": "numeric",
            },
          ]
        }
        data={Array []}
        icons={
          Object {
            "Add": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Check": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Clear": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Delete": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "DetailPanel": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Edit": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Export": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Filter": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "FirstPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "LastPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "NextPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "PreviousPage": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ResetSearch": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "Search": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "SortArrow": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ThirdStateCheck": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
            "ViewColumn": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        options={
          Object {
            "filtering": true,
          }
        }
        title="Report"
      />
    </ForwardRef(Box)>
  </ForwardRef(Stack)>
</ForwardRef(Box)>
`;
